src/jdk.jdwp.agent/share/native/libjdwp/debugLoop.c

```diff
@@ -87,7 +87,9 @@ debugLoop_run(void)
     /* Initialize all statics */
     /* We may be starting a new connection after an error */
     cmdQueue = NULL;
-    cmdQueueLock = debugMonitorCreate("JDWP Command Queue Lock");
+    if (cmdQueueLock == NULL) {
```

Comment:
Doesn't it makes sense to check that lock is not locked here?
Is there possibility that lock is left locked for disconnected process or any other similar timeout. Probably it is not covered by our tests.

Reply:
The normal status of the cmdQueueLock is for the Listener thread to be waiting on it. When the Reader thread gets a packet and queues it up, it then does a notify. Any holding of the lock is short lived. I don't see how it could still be locked at this point, and if it was, it should be quickly released. I typed up the following notes for myself a while back when figuring this all out. They may be of some help to you in understanding whe this should not be a problem.

=========================================

The debug agent has two threads involved in dealing with the connection to the debugger. The first is the JDWP Transport Listener thread. It is initially created to listen for the debugger's connection request. Once the request comes in the connection is setup and the Listener thread spawns off the JDWP Command Reader thread. The Reader thread is responsible for reading in command packets and queuing them up for the Listener thread to process.

The Reader thread is usually waiting for the next JDWP command, which it will enqueue when it arrives:
```
    while (shouldListen) {
        jint rc;

        rc = transport_receivePacket(&packet);
        ...
        enqueue(&packet);  <-- does a notify on cmdQueueLock
        ...
    }
```

The Listener thread sits in a loop dequeuing commands until it gets a VM.Dispose command, at which point it will exit the loop, destroy the cmdQueueLock, and then exit the thread. Usually while dequeuing it is blocked here waiting for the next command:

```
    debugMonitorEnter(cmdQueueLock);
    while (!transportError && (cmdQueue == NULL)) {
        debugMonitorWait(cmdQueueLock);
    }
    debugMonitorExit(cmdQueueLock);
```

It will process enqueued packets in this manner until it gets a VM.Dispose command.

The cmdQueueLock is created by the Listener thread just before spawning the Reader thread. It is destroyed after it exits as described above. If VM.Dispose is being used, things are pretty orderly, but we have to also consider abnormal connection termination where no VM.Dispose is done. In that case the the above transport_receivePacket() call will fail and the Reader thread calls the following code:

```
    debugMonitorEnter(cmdQueueLock);
    transportError = JNI_TRUE;
    debugMonitorNotify(cmdQueueLock);
    debugMonitorExit(cmdQueueLock);
```

The notify wakes up the Listener thread, which is in the debugMonitorWait() trying to dequeue the next command. The dequeue fails, which results in the Listener thread shutting down the connection. It will destroy the cmdQueueLock just like it does in the orderly shutdown.